### ISSUE TITLE
[DocDB] Read requests could be stuck in the wait-queue temporarily when a tablet receives concurrent reads for the transaction

### ISSUE THREAD

**Issue by basavaraj29** [yugabyte/yugabyte-db#22047]
Jira Link: [DB-10964](https://yugabyte.atlassian.net/browse/DB-10964)
### Description

The wait-queue tracks wait-for dependencies on a rpc request level basis. We use the retryable request id to differentiate between concurrent requests of the same transaction.

Only 2 types of reads would launch `WriteQuery` object, and end up entering the wait-queue. 
1. Serializable reads
2. reads with explicit locks

Since reads rpc(s) don't have a retryable request id, we populate the request id as -1 for the launched `WriteQuery` during conflict resolution.
```
auto request_id = request().has_request_id() ? request().request_id() : -1;
```

Each blocker stores its waiters on a request id granularity. When the blocker resolves, we try to signal the waiter request. 

In one of the stress runs, we see the following error 
```
E0417 20:26:22.514000 31926 wait_queue.cc:548] TxnId: de4075c6-2ef4-40b6-8530-7c83577b0534 Found multiple waiters with the same request_id -1
```
It implies that there were multiple concurrent read requests of a txn at one of the tablets. Since we only resume the request with latest timestamp, the other requests with id -1 of the txn would only be resumed when they timeout in the wait-queue (every 30 secs).

A potential fix would be to resume all requests with the same id (in this case, -1), but we need to investigate what reads have this behavior. And it is definitely not expected to see multiple requests of a txn with the same id other than -1.

### Issue Type

kind/bug

### Warning: Please confirm that this issue does not contain any sensitive information

- [X] I confirm this issue does not contain any sensitive information.

[DB-10964]: https://yugabyte.atlassian.net/browse/DB-10964?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by robertsami** [comment]
@basavaraj29 i think something like this should resolve this issue:

```
diff --git a/src/yb/docdb/conflict_resolution.cc b/src/yb/docdb/conflict_resolution.cc
index d3719d73815..ba4bcd669b6 100644
--- a/src/yb/docdb/conflict_resolution.cc
+++ b/src/yb/docdb/conflict_resolution.cc
@@ -669,8 +669,8 @@ class WaitOnConflictResolver : public ConflictResolver {
     DCHECK(!status_tablet_id_.empty());
     auto did_wait_or_status = wait_queue_->MaybeWaitOnLocks(
         context_->transaction_id(), context_->subtransaction_id(), lock_batch_, status_tablet_id_,
-        serial_no_, context_->GetTxnStartUs(), request_start_us_, request_id_, deadline_,
-        std::bind(&WaitOnConflictResolver::GetLockStatusInfo, shared_from(this)),
+        serial_no_, context_->GetTxnStartUs(), request_start_us_, GetRequestIdForWaitQueue(),
+        deadline_, std::bind(&WaitOnConflictResolver::GetLockStatusInfo, shared_from(this)),
         std::bind(&WaitOnConflictResolver::WaitingDone, shared_from(this), _1, _2));
     if (!did_wait_or_status.ok()) {
       InvokeCallback(did_wait_or_status.status());
@@ -691,7 +691,7 @@ class WaitOnConflictResolver : public ConflictResolver {
     return wait_queue_->WaitOn(
         context_->transaction_id(), context_->subtransaction_id(), lock_batch_,
         ConsumeTransactionDataAndReset(), status_tablet_id_, serial_no_,
-        context_->GetTxnStartUs(), request_start_us_, request_id_, deadline_,
+        context_->GetTxnStartUs(), request_start_us_, GetRequestIdForWaitQueue(), deadline_,
         std::bind(&WaitOnConflictResolver::GetLockStatusInfo, shared_from(this)),
         std::bind(&WaitOnConflictResolver::WaitingDone, shared_from(this), _1, _2));
   }
@@ -726,6 +726,10 @@ class WaitOnConflictResolver : public ConflictResolver {
   }
 
  private:
+  int64_t GetRequestIdForWaitQueue() const {
+    return request_id_ > 0 ? request_id_ : serial_no_;
+  }
+
   WaitQueue* const wait_queue_;
   LockBatch* const lock_batch_;
   uint64_t serial_no_;

```

this probably should have been included in 2896530

**Comment by rthallamko3** [comment]
Reactivating for 2024.1.0 backport
